samples/matter/common/src/binding_handler.cpp

```diff
@@ -0,0 +1,156 @@
+
+		if (CHIP_NO_ERROR != error) {
+			LOG_ERR("NotifyBoundClusterChanged failed due to: %" CHIP_ERROR_FORMAT, error.Format());
+		}
+	} else {
+		Platform::Delete<BindingData>(bindingData);
+		LOG_ERR("Binding command was not applied! Reason: %" CHIP_ERROR_FORMAT, Error.Format());
+	}
+}
+
+void BindingHandler::DeviceChangedCallback(const EmberBindingTableEntry &binding, OperationalDeviceProxy *deviceProxy,
+					   void *context)
+{
+	VerifyOrReturn(context != nullptr, LOG_ERR("Invalid context for device handler"));
+	BindingData *data = static_cast<BindingData *>(context);
+
+	if (binding.type == EMBER_MULTICAST_BINDING && data->IsGroup) {
+		data->InvokeCommandFunc(binding, nullptr, *data);
+	} else if (binding.type == EMBER_UNICAST_BINDING && !(data->IsGroup)) {
+		data->InvokeCommandFunc(binding, deviceProxy, *data);
+	}
+}
+
+void BindingHandler::DeviceContextReleaseHandler(void *context)
+{
+	VerifyOrDie(context != 0);
+
+	Platform::Delete(static_cast<BindingData *>(context));
+}
+
+void BindingHandler::InitInternal()
+{
+	LOG_INF("Initialize binding Handler");
+	auto &server = Server::GetInstance();
+	if (CHIP_NO_ERROR !=
+	    BindingManager::GetInstance().Init(
+		    { &server.GetFabricTable(), server.GetCASESessionManager(), &server.GetPersistentStorage() })) {
+		LOG_ERR("BindingHandler::InitInternal failed");
+	}
+
+	BindingManager::GetInstance().RegisterBoundDeviceChangedHandler(DeviceChangedCallback);
+	BindingManager::GetInstance().RegisterBoundDeviceContextReleaseHandler(DeviceContextReleaseHandler);
+	BindingHandler::PrintBindingTable();
+}
+
+bool BindingHandler::IsGroupBound()
+{
+	BindingTable &bindingTable = BindingTable::GetInstance();
+
+	for (auto &entry : bindingTable) {
+		if (EMBER_MULTICAST_BINDING == entry.type) {
+			return true;
+		}
+	}
+	return false;
+}
+
+void BindingHandler::PrintBindingTable()
+{
+	BindingTable &bindingTable = BindingTable::GetInstance();
+
+	LOG_INF("Binding Table size: [%d]:", bindingTable.Size());
+	uint8_t i = 0;
+	for (auto &entry : bindingTable) {
+		switch (entry.type) {
+		case EMBER_UNICAST_BINDING:
+			LOG_INF("[%d] UNICAST:", i++);
+			LOG_INF("\t\t+ Fabric: %d\n \
+            \t+ LocalEndpoint %d \n \
+            \t+ ClusterId %d \n \
+            \t+ RemoteEndpointId %d \n \
+            \t+ NodeId %d",
+				(int)entry.fabricIndex, (int)entry.local, (int)entry.clusterId.Value(),
+				(int)entry.remote, (int)entry.nodeId);
+			break;
+		case EMBER_MULTICAST_BINDING:
+			LOG_INF("[%d] GROUP:", i++);
+			LOG_INF("\t\t+ Fabric: %d\n \
+            \t+ LocalEndpoint %d \n \
+            \t+ RemoteEndpointId %d \n \
+            \t+ GroupId %d",
+				(int)entry.fabricIndex, (int)entry.local, (int)entry.remote, (int)entry.groupId);
+			break;
+		case EMBER_UNUSED_BINDING:
+			LOG_INF("[%d] UNUSED", i++);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+void BindingHandler::DeviceWorkerHandler(intptr_t context)
+{
+	VerifyOrDie(context != 0);
```

Comment:
Isn't it too aggressive? Should be really kill the app? BTW. Isn't is better to check for the nullptr after line 147?

Reply:
This is not passed by a class user, so if this ends up nullptr it means this component is broken so I would say assertions/VerifyOrDie are exactly for these situations :)

Reply:
Fair enough :)